src/main/java/org/skriptlang/skript/lang/errors/RuntimeErrorProducer.java

```diff
@@ -0,0 +1,158 @@
+package org.skriptlang.skript.lang.errors;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.config.Config;
+import ch.njol.skript.config.Node;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.lang.Condition;
+import ch.njol.skript.lang.Effect;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.Section;
+import ch.njol.skript.lang.SkriptParser;
+import ch.njol.skript.lang.SyntaxElement;
+import ch.njol.skript.lang.parser.ParserInstance;
+import ch.njol.skript.localization.ArgsMessage;
+import ch.njol.skript.util.Utils;
+import ch.njol.util.Kleenean;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.skriptlang.skript.lang.structure.Structure;
+
+import java.util.logging.Level;
+
+/**
+ * A RuntimeErrorProducer can throw runtime errors in a standardized and controlled manner.
+ * Only {@link SyntaxElement}s are intended to implement this interface.
+ */
+public interface RuntimeErrorProducer {
+
+	// todo: try/catch, suppression when annotations are added
+
+	/**
+	 * Returns the {@link Node} that this is a part of. Used for accessing the line contents via {@link Node#getKey()}
+	 * and the line number via {@link Node#getLine()}.
+	 * <br>
+	 * A standard implementation is to store the Node during
+	 * {@link SyntaxElement#init(Expression[], int, Kleenean, SkriptParser.ParseResult)}
+	 * via {@link ParserInstance#getNode()}.
+	 * @return The node that produced a runtime error.
+	 */
+	Node getNode();
+
+	/**
+	 * Gets the text that should be underlined within the line contents. This should match the text the user wrote that
+	 * was parsed as the syntax that threw the runtime issue. For example, if the skull expression in
+	 * {@code give skull of player to all players} throws a runtime error, this method should return
+	 * {@code "skull of player"}
+	 * <br>
+	 * An example implementation for {@link Expression}s is to store {@link SkriptParser.ParseResult#expr} during
+	 * {@link SyntaxElement#init(Expression[], int, Kleenean, SkriptParser.ParseResult)} and return that.
+	 * <br>
+	 * For other syntax types, this may vary. Effects, for example, may underline the whole line.
+	 *
+	 * @return The text to underline in the line that produced a runtime error.
+	 */
+	String toUnderline();
```

Comment:
Do you think a name like `toHighlight` might work better? Since an underline is not necessarily applicable everywhere.